### PR TITLE
cmd/snap-repair: filter repair assertions based on bases + modes

### PR DESCRIPTION
**Pull request by anonymouse64**
Use the bases and modes settings in repair assertions to filter repair
assertions when snap-repair is running.

The logic for modes is such that when modes is not set, on UC20 devices,
the assertion only applies to run mode devices. If modes include "run",
non-UC20 devices will also be included. To filter just on UC20 or UC18
devices, the bases setting should be set in assertion.

A spread test using test keys, etc. will be a followup.

### CODE REVIEW

File: cmd/snap-repair/runner.go [hunk 1]
```diff
@@ -900,8 +906,50 @@ func (run *Runner) Applicable(headers map[string]interface{}) bool {
...
+	}
+
+	// modes is slightly more nuanced, if the modes setting in the assertion
+	// header is unset, that means that the assertion should only be run in
+	// "run" mode on uc20
+
+	if run.state.Device.Mode == "" {
+		// uc16 / uc18 device, the assertion is only applicable to us if modes
+		// has "run" in it or if it is unset
+		if len(modes) != 0 && !strutil.ListContains(modes, "run") {
```
Comment (pedronis) [inline cmd/snap-repair/runner.go:931]: modes in the assertion can only be non-empty if bases are core2* or empty, so the only relevant case would be a repair with:

```
modes: [run]
(no bases)
```

I think such a repair should also imply core20+ only

Comment (anonymouse64) [reply cmd/snap-repair/runner.go:931]: how do we handle the implication of core20+ here? do we want to hard-code a block list of core and core18, and assume all other base snaps are core20+ like we did previously?

Comment (pedronis) [reply cmd/snap-repair/runner.go:931]: my naive thinking was to just have:

```
if len(modes) != 0 {
    return false
}
```

on this path where `Device.Mode == ""`

Comment (anonymouse64) [reply cmd/snap-repair/runner.go]: this is done now

File: cmd/snap-repair/runner.go [hunk 2]
```diff
@@ -900,8 +906,48 @@ func (run *Runner) Applicable(headers map[string]interface{}) bool {
...
+	}
+
+	modes, err := stringList(headers, "modes")
+	if err != nil {
+		return false
+	}
+
+	// modes is slightly more nuanced, if the modes setting in the assertion
+	// header is unset, that means that the assertion should only be run in
+	// "run" mode on uc20
```
Comment (pedronis) [inline cmd/snap-repair/runner.go:926]: this comment sounds wrong or I'm misunderstanding it

**Review by pedronis — APPROVED** [review]
thanks, one comment seems off though

File: cmd/snap-repair/runner.go (hunk 2 above)
Comment (anonymouse64) [reply cmd/snap-repair/runner.go:926]: yeah this is confusing, I will rephrase

File: cmd/snap-repair/runner_test.go [hunk 3]
```diff
@@ -1346,6 +1433,348 @@ AXNpZw==`}
...
+
+AXNpZw==
+	`
+
+	r1 := sysdb.InjectTrusted(s.storeSigning.Trusted)
+	defer r1()
+	r2 := repair.MockTrustedRepairRootKeys([]*asserts.AccountKey{s.repairRootAcctKey})
+	defer r2()
+
+	tt := []struct {
```
Comment (mvo5) [inline cmd/snap-repair/runner_test.go:1458]: Thanks for these tests!

**Review by mvo5 — APPROVED** [review]

**Pull request merged**